### DISCUSSION TITLE
Where to find a working full example of how to migrate from standard SQLalchemy to FastApi-Users

### DISCUSSION THREAD

**Discussion by sorasful** [fastapi-users/fastapi-users#812] [Q&A]
Hello there ! 

I'm currently in the process of setting up authentication on my fastapi app. For now I went from scratch and wanted to use this package since it seems to come with a lot of great features ! 

The only problem I have, is that I can't really find full examples on how to make this work with some custom User model. I saw the documentation about adding an additionnal field, but I didn't get a good idea on how to migrate from a SQLalchemy fastapi app, to using fastapi-users.

I also searched for tutorials or repositories, but not much luck here. 

The idea would to go from very basic and classic model like this : 

```python
class User(Base):
    __tablename__ = 'users'

    id: int = Column(Integer, primary_key=True)
    email: str = Column(String, nullable=False)
    username: str = Column(String, nullable=False)
    creation_date: datetime = Column(DateTime(True), nullable=False, server_default=func.now())

```

And go to setting up fastapi-users. I think it would also attrack people if there were a "migrate from XXX" section in the documentation.

Anyway, thanks very much for your time and help !

**Comment by frankie567** [comment]
Hi there!

The thing is in this case, there is a much migration path as there are people, since everyone implemented it with its own ideas 🙃

In your case, the migration will be a bit complicated for two reasons:

* You use a classic auto-incremented integer id. In FastAPI Users, the id is a UUID4.
* You have a username column. In FastAPI Users, we don't have the notion of username: only the e-mail is used to login. You can keep a username, of course, but you won't be able to use it as a login.

If you already run your application in production with real data, well, you'll have to write migration scripts to change the users id (and all the foreign keys pointing to it) to UUID4.

If your application is still in development, well, it might be better to just replace all the users management with the standard installation of FastAPI Users.

Adding your own fields is pretty easy. The [basic SQLAlchemy installation](https://fastapi-users.github.io/fastapi-users/configuration/databases/sqlalchemy/#create-the-tables) looks like the following:

```
class UserTable(Base, SQLAlchemyBaseUserTable):
    pass
```

This is an SQLAlchemy table. `SQLAlchemyBaseUserTable` provides all the basic fields for FastAPI Users. You can then add your own like this:

```
class UserTable(Base, SQLAlchemyBaseUserTable):
    username: str = Column(String, nullable=False)
    creation_date: datetime = Column(DateTime(True), nullable=False, server_default=func.now())
```

Don't forget also to mirror this in the Pydantic models as explained [here](https://fastapi-users.github.io/fastapi-users/configuration/models/#adding-your-own-fields).

**Comment by sorasful** [comment]
Hey ! Thank you for your quick reply ! 

You're right, I'm still under development so the migration would be easy enough, but, about these points : 

> You use a classic auto-incremented integer id. In FastAPI Users, the id is a UUID4.  

That's easy to change, so no really big troubles here.

> You have a username column. In FastAPI Users, we don't have the notion of username: only the e-mail is used to login. You can keep a username, of course, but you won't be able to use it as a login.

Hmm, I see. Is this something that would be interresting to add in fastapi-users ? Being able to log in using email + others fields if specified. If so, I can try to see if that would be possible and try to do a PR for this. 


Also, I have just a question about these Pydantic models taken from the example : 


```python
import datetime

from fastapi_users import models


class User(models.BaseUser):
    first_name: str
    birthdate: Optional[datetime.date]


class UserCreate(models.BaseUserCreate):
    first_name: str
    birthdate: Optional[datetime.date]


class UserUpdate(models.BaseUserUpdate):
    first_name: Optional[str]
    birthdate: Optional[datetime.date]


class UserDB(User, models.BaseUserDB):
    pass

```


I don't really understand the role of UserDB, I read that it's basically the representation of the data in db, but you add a hashed password field. But when does one needs to update this UserDB model ? It's a little bit unclear to me.


Thanks for your time :)
